### PR TITLE
DAPM graph fixes and clean ups

### PR DESCRIPTION
**Pull request by ranj063**
This PR does the following:
1. Fixes issue #661 
2. removes the duplicate connections in the dapm graph from the PCM dev to host comp
3. cleans up the dapm graph by removing the dai capabilities from the dai definitions

### CODE REVIEW

**Comment by ranj063** [comment]
This PR should be merged alongside https://github.com/thesofproject/sof/pull/1047

**Comment by ranj063** [comment]
this PR also fixes https://github.com/thesofproject/linux/issues/661 

Confirmed by Yong to fix the noise issue on the chromebook. @keqiaozhang can you please confirm?

**Comment by ranj063** [comment]
removing the DAI capabilities from the dai definitions cleans up the dapm graph.
**Before:**
pcm512x-idle

**After**
pcm-512x-cleanup

File: sound/soc/sof/topology.c [hunk 1]
```diff
@@ -2031,6 +2059,15 @@ static int sof_dai_load(struct snd_soc_component *scomp, int index,
 		goto free_playback_tables;
 	}
 
+	/* bind pcm to host comp */
+	ret = spcm_bind(sdev, spcm, stream);
+	if (ret) {
+		dev_err(sdev->dev,
+			"error: can't bind pcm to host\n");
+		snd_dma_free_pages(&spcm->stream[stream].page_table);
+		goto free_playback_tables;
```
Comment (bardliao) [inline sound/soc/sof/topology.c:2068]: Should we also free capture tables here?

Comment (ranj063) [reply sound/soc/sof/topology.c:2068]: @bardliao 		snd_dma_free_pages(&spcm->stream[stream].page_table); frees the capture tables before freeing the playback tables

Comment (bardliao) [reply sound/soc/sof/topology.c:2068]: @ranj063 Oh, yes. Didn't notice there is a snd_dma_free_pages before the goto statement :)

File: sound/soc/soc-dapm.c [hunk 2]
```diff
@@ -3651,6 +3651,8 @@ snd_soc_dapm_new_control_unlocked(struct snd_soc_dapm_context *dapm,
 	case snd_soc_dapm_dac:
 	case snd_soc_dapm_aif_in:
 	case snd_soc_dapm_pga:
+	case snd_soc_dapm_buffer:
+	case snd_soc_dapm_scheduler:
```
Comment (plbossart) [inline sound/soc/soc-dapm.c:3655]: what about all the effects/src/asrc that were added at the same time as buffer/scheduler? Should they really be always on?

File: sound/soc/sof/intel/bdw.c [hunk 3]
```diff
@@ -640,24 +640,13 @@ static int bdw_probe(struct snd_sof_dev *sdev)
 	return ret;
 }
 
-#define BDW_FORMATS (SNDRV_PCM_FMTBIT_S16_LE | SNDRV_PCM_FMTBIT_S24_LE | \
-	SNDRV_PCM_FMTBIT_S32_LE)
-
```
Comment (plbossart) [inline sound/soc/sof/intel/bdw.c:645]: these formats may not be used, but at the same time we currently have restrictions on what the topology can do (single format per DAI), and there will be no check that the topology settings are correct and supported by the firmware. We should probably have a layer that does check the topology is indeed supported, otherwise with a single error in a topology file we have a risk of reaching unsupported configurations. it's just too risky.
In short, I would really like the topology to support multiple DAI configs/formats, and a per-platform check in the driver.

Comment (ranj063) [reply sound/soc/sof/intel/bdw.c:645]: @plbossart makes sense. I can add the check but looking at the formats defined for all platforms, I see no restrictions. All platforms seem to support 16/24/32 bit formats. Is that a mistake? If it is, could you please remind me what they should look like?

**Comment by plbossart** [comment]
> -#define BDW_FORMATS (SNDRV_PCM_FMTBIT_S16_LE | SNDRV_PCM_FMTBIT_S24_LE | \
> -	SNDRV_PCM_FMTBIT_S32_LE)
> -
>
> @plbossart <https://github.com/plbossart> makes sense. I can add the 
> check but looking at the formats defined for all platforms, I see no 
> restrictions. All platforms seem to support 16/24/32 bit formats. Is 
> that a mistake? If it is, could you please remind me what they should 
> look like?
>
the main goal is to make sure the SSP are configured in topology with 
1..8 channels, frequency between 8..192 and with 16, 24, 32 bits. There 
are other exotic formats such as 24_3LE or FLOAT that should not be 
handled. More channels as well.

**Comment by ranj063** [comment]
> -#define BDW_FORMATS (SNDRV_PCM_FMTBIT_S16_LE | SNDRV_PCM_FMTBIT_S24_LE | \ -	SNDRV_PCM_FMTBIT_S32_LE) - @plbossart <https://github.com/plbossart> makes sense. I can add the check but looking at the formats defined for all platforms, I see no restrictions. All platforms seem to support 16/24/32 bit formats. Is that a mistake? If it is, could you please remind me what they should look like?
> the main goal is to make sure the SSP are configured in topology with 1..8 channels, frequency between 8..192 and with 16, 24, 32 bits. There are other exotic formats such as 24_3LE or FLOAT that should not be handled. More channels as well.

@plbossart another question. Is sof_pcm_dai_link_fixup() the place to do the check? Today we seem to align the dai format and channels in the fixup method to fix the min and max rate to fsync rate and the min and max channels to tdm_slots. Do these need to expand to support a range?

File: sound/soc/soc-dapm.c (hunk 2 above)
Comment (ranj063) [reply sound/soc/soc-dapm.c:3655]: @plbossart no, added them too now.

**Comment by ranj063** [comment]
@plbossart I've updated the PR with the check for SSP fsync rate and channel count.

File: sound/soc/soc-dapm.c [hunk 4]
```diff
@@ -3651,6 +3651,10 @@ snd_soc_dapm_new_control_unlocked(struct snd_soc_dapm_context *dapm,
...
+	case snd_soc_dapm_src:
+	case snd_soc_dapm_effect:
```
Comment (plbossart) [inline sound/soc/soc-dapm.c:3656]: Meh. this isn't the same order as definitions and it's missing asrc/encoder/decoder.

**Comment by ranj063** [comment]
> Minor issue with the soc-dapm correction, we're missing asrc/encoder/decoder.

oh yes, fixed now @plbossart
